### PR TITLE
test: integration tests for akri and opcua checks

### PR DESCRIPTION
**Pull request by vilit1**
Akri check integration tests

1. organize check integration tests
2. set up helpers for other service check tests (opcua, mq)
3. see that `--as-object` needs love

image

Future tasks:
1. improve akri checks --as-object
2. improve opcua checks --as-object
3. add mq check int tests
5. fill out pre check int test

---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to Azure IoT Operations tooling!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT Operations CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

### CODE REVIEW

File: azext_edge/tests/edge/checks/int/helpers.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    if not resource_kind_present:
+        assert key is None
+        return
+    assert post_deployment[key]
+    assert post_deployment[key]["description"].startswith(f"Evaluate {description_name}")
+    # for the ones that have spaces
+    assert post_deployment[key]["description"].replace(" ", "").endswith(resource_plural)
+
+    # check the target exisitance
```
Comment (Elsie4ever) [inline azext_edge/tests/edge/checks/int/helpers.py:114]: nit: existence

File: azext_edge/tests/helpers.py [hunk 2]
```diff
@@ -7,13 +7,115 @@
...
+    def pass_conditions(item: Dict[str, Any]) -> bool:
+        name = item["metadata"]["name"]
+        if resource_match and not fnmatch(name, resource_match):
+            return False
+        if prefixes and not any(name.startswith(prefix) for prefix in prefixes):
+            return False
+        return True
+
+    filtered = {}
+    for item in kubectl_items["items"]:
```
[the remark this answers is not in the document]
Comment (Elsie4ever) [reply azext_edge/tests/helpers.py:35]: adding if check for kubectl_items.get("items")?

**Comment by vilit1** [comment]
new cluster, `--include-dp` args
image

File: azext_edge/tests/edge/checks/int/test_akri_int.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+from ....helpers import get_kubectl_custom_items
+from ....generators import generate_names
+
+logger = get_logger(__name__)
+
+
+@pytest.mark.parametrize("detail_level", ResourceOutputDetailLevel.list())
+@pytest.mark.parametrize("resource_kind", AkriResourceKinds.list() + [None])
+# TODO: figure out if name match should be a general test vs each service (minimize test runs)
+@pytest.mark.parametrize("resource_match", [None, "*otel*", "akri-opcua*", generate_names()])
```
Comment (vilit1) [inline azext_edge/tests/edge/checks/int/test_akri_int.py:29]: leaning towards removing resource match testing here (and in opcua) and adding a todo for general resource match testing.

File: azext_edge/tests/helpers.py [hunk 4]
```diff
@@ -7,13 +7,115 @@
...
 
 
+def filter_resources(
+    kubectl_items: List[Dict[str, Any]],
+    prefixes: Optional[Union[str, List[str]]] = None,
+    resource_match: Optional[str] = None,
+) -> Dict[str, Any]:
+    def pass_conditions(item: Dict[str, Any]) -> bool:
+        name = item["metadata"]["name"]
+        if resource_match and not fnmatch(name, resource_match):
```
Comment (vilit1) [inline azext_edge/tests/helpers.py:28]: can change into a single return

`return any([
   not resource_match and not prefixes,
   resource_match and fnmatch(name, resource_match),
   prefixes and any(name.startswith(prefix) for prefix in prefixes)
])`

File: azext_edge/tests/edge/checks/int/helpers.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+):
+    assert post_deployment["evalCoreServiceRuntime"]
+    assert post_deployment["evalCoreServiceRuntime"]["description"] == f"Evaluate {description_name} core service"
+    overall_status = "success"
+    runtime_resource = post_deployment["evalCoreServiceRuntime"]["targets"]["coreServiceRuntimeResource"]
+    for namespace in runtime_resource.keys():
+        namespace_status = "success"
+        assert not runtime_resource[namespace]["conditions"]
+        evals = runtime_resource[namespace]["evaluations"]
+        kubectl_pods = get_kubectl_workload_items(
```
Comment (vilit1) [inline azext_edge/tests/edge/checks/int/helpers.py:56]: will need to make this into a helper function for mq pod checks

File: azext_edge/tests/edge/checks/int/helpers.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    if resource_kind:
+        command += f"--resources {resource_kind} "
+    if resource_match:
+        command += f"--resource-name {resource_match} "
+    result = run(command)
+
+    return {cond["name"]: cond for cond in result["postDeployment"]}, service_present
+
+
+def sort_by_namespace(
```
Comment (vilit1) [inline azext_edge/tests/edge/checks/int/helpers.py:154]: I should add in doc strings describing what the functions do...

File: azext_edge/tests/helpers.py [hunk 7]
```diff
@@ -7,13 +7,115 @@
...
+    service_type: str,
+    namespace: Optional[str] = None,
+    resource_match: Optional[str] = None,
+) -> Dict[str, Any]:
+    if service_type == "pvc":
+        service_type = "persistentvolumeclaim"
+    if isinstance(prefixes, str):
+        prefixes = [prefixes]
+    namespace_param = f"-n {namespace}" if namespace else "-A"
+    kubectl_items = run(f"kubectl get {service_type}s {namespace_param} -o json")
```
Comment (c-ryan-k) [inline azext_edge/tests/helpers.py:111]: A word of caution for use with future checks - I believe certain CRDs may have a plural that isn't `s`

**Pull request merged**